File: Python/FourNumbersSum.py
#average: O(n^2) time | O(n^2) space
#Worst: O(n^3) time | O(n^2) space
def fourNumberSum(array, targetSum):
    allPairSum = {}
    quadruplets = []
    import ipdb
    ipdb.set_trace()
    for i in range(1, len(array) - 1):
        print(i)
        for j in range(i + 1, len(array)):
            print(j)
            currentSum = array[i] + array[j]
            difference = targetSum - currentSum
            if difference in allPairSum:
                for pair in allPairSum[difference]:
                    quadruplets.append(pair + [array[i], array[j]])
        for k in range(0, i):
            currentSum = array[i] + array[k]
            if currentSum not in allPairSum:
                allPairSum[currentSum] = [[array[k], array[i]]]
            else:
                allPairSum[currentSum].append([array[k], array[i]])
    print(quadruplets)
    return quadruplets

array = [1,3,5,7,9,12,33]

fourNumberSum(array,61)

File: Java/FourNumbersSum.java
import java.util.*;
/**
 * <h1> Four Numbers Sum </h1>
 * <p> Algoritmo para obtener 4 digitos que sumen un resultado dado </p>
 * <p>  </p>
 * { @code fourNumbersum() }
 * @author Roman Vargas
 * @since 1.0
 * @version 1.0
 */
class FourNumbersSum {
  
  /** 
   * @param array
   * @param targetSum
   * @return List<Integer[]>
   */
  public static List<Integer[]> fourNumberSum(int[] array, int targetSum) {
    Map<Integer, List<Integer[]>> allPairSums = new HashMap<>();
		List<Integer[]> quadruplets = new ArrayList<Integer[]>();
		for(int i = 1; i < array.length - 1; i++) {
			for(int j = i + 1; j < array.length; j++) {
				int currentSum = array[i] + array[j];
				int difference = targetSum - currentSum;
				if(allPairSums.containsKey(difference)) {
					for(Integer[] pair: allPairSums.get(difference)) {
						Integer[] newQuadruplet = {pair[0], pair[1], array[i], array[j]};
						quadruplets.add(newQuadruplet);
					}
				}
			}
			for(int k = 0; k < i; k++){
				int currentSum = array[i] + array[k];
				Integer[] pair = {array[k], array[i]};
				if(!allPairSums.containsKey(currentSum)) {
					List<Integer[]> pairsGroup = new ArrayList<Integer[]>();
					pairsGroup.add(pair);
					allPairSums.put(currentSum, pairsGroup);
				} else {
					allPairSums.get(currentSum).add(pair);
				}
			}
		}
    return quadruplets;
  }
}